buffer/src/main/java/io/netty/buffer/api/BufferAllocator.java

```diff
@@ -0,0 +1,170 @@
+     * @return The newly allocated {@link Buffer}.
+     */
+    default Buffer allocate(int size, ByteOrder order) {
+        return allocate(size).order(order);
+    }
+
+    /**
+     * Create a supplier of "constant" {@linkplain Buffer Buffers} from this allocator, that all have the given
+     * byte contents. The buffer has the same capacity as the byte array length, and its write offset is placed at the
+     * end, and its read offset is at the beginning, such that the entire buffer contents are readable.
+     * <p>
+     * The buffers produced by the supplier will each have their own independent life-cycle, and closing them will
+     * make them {@linkplain Buffer#isAccessible() inaccessible}, just like normally allocated buffers.
+     * <p>
+     * The buffers produced are "constants", in the sense that they are {@linkplain Buffer#readOnly() read-only}.
+     * <p>
+     * It can generally be expected, but is not guaranteed, that the returned supplier is more resource efficient than
+     * allocating and copying memory with other available APIs. In such optimised implementations, the underlying memory
+     * baking the buffers will be shared among all the buffers produced by the supplier.
+     * <p>
+     * The primary use case for this API, is when you need to repeatedly produce buffers with the same contents, and
+     * you perhaps wish to keep a {@code static final} field with these contents. This use case has previously been
+     * solved by allocating a read-only buffer with the given contents, and then slicing or duplicating it on every use.
+     * This approach had several problems. For instance, if you forget to slice, the offsets of the buffer can change
+     * in unexpected ways, since the same buffer instance is shared and accessed from many places. The buffer could also
+     * be deallocated, making the data inaccessible. The supplier-based API solves all of these problems, by enforcing
+     * that each usage get their own distinct buffer instance.
+     *
+     * @param bytes The byte contents of the buffers produced by the returned supplier.
+     * @return A supplier of read-only buffers with the given contents.
+     */
+    default Supplier<Buffer> constBufferSupplier(byte[] bytes) {
+        byte[] safeCopy = bytes.clone(); // Prevent modifying the bytes after creating the supplier.
+        return () -> allocate(bytes.length).writeBytes(safeCopy).makeReadOnly();
+    }
+
+    /**
+     * Close this allocator, freeing all of its internal resources.
+     * <p>
+     * Existing (currently in-use) allocated buffers will not be impacted by calling this method.
+     * If this is a pooling or caching allocator, then existing buffers will be immediately freed when they are closed,
+     * instead of being pooled or cached.
+     * <p>
+     * The allocator can still be used to allocate more buffers after calling this method.
```

Comment:
The shut-down process of systems tend to have high bug density as it's generally hard to transition a system, bit by bit, from steady-state, to terminated. An allocator that allows allocations after it has been shut down, can be shut down out of order with the rest of the system, without causing catastrophic failures in other unknown parts of the system.

It may look weird, but I think this will be helpful to a lot of people in practice.

Reply:
I am not sure, I think relaxed behaviors like this are more confusing than fail-fast and expect folks to close in order. Allowing usage post `close()` is counter-intuitive specifically when allowed for a subset of contracts.

Reply:
> Close this allocator, freeing all of its internal resources.

Should this be named `startClose()`  if it doesn't complete after return?